ddev/ddev
# feat: expand `.ddev/.env.*` files on `ddev start`, add `ddev dotenv`, add variable expansion to `ddev add-on get`, fixes #4430

## The Issue

- #4430

## How This PR Solves The Issue

* Previously, variables from the `.env` file were expanded at the `docker-compose config` build stage for all `docker-compose.*.yaml` files, but were passed only to the `web` container. And the documentation says another thing:

	> An optional, project-level `.ddev/.env` file provides environment variables to all DDEV containers, including any additional services or add-ons.

	This PR changes the behavior to pass all variables from `.ddev/.env` to all containers.

* Adds support for custom long flags with `ddev dotenv set`, which allows to do things like:

	```
	ddev dotenv set .ddev/.env.redis --test-variable=foo --some-addon-value 123 --enable-debug --test-variable2 test
	```
	which adds `.ddev/.env.redis` with:
	```dotenv
	TEST_VARIABLE="foo"
	SOME_ADDON_VALUE="123"
	ENABLE_DEBUG=""
	TEST_VARIABLE2="test"
	```
	As you can see above, `--enable-debug` is converted to an empty value, which makes it easier for add-on maintainers to add simple flags without values, to check the existence of a variable instead of checking the value.

	Uppercase flags or short flags will not work, the below flags will be ignored:
	```
	ddev dotenv set .ddev/.env.redis --TEST=123 -a -B 123 -test=foo
	```

	These variables can be read with `ddev dotenv get .ddev/.env.redis --variable-name`.

	This can be useful in bash scripts:
	```bash
    TEST_VARIABLE=$(ddev dotenv get .ddev/.env.redis --test-variable 2>/dev/null)
    exit_code=$?
    if [[ $exit_code -eq 0 && "${TEST_VARIABLE}" == "foo" ]]; then
        # do something
    fi

* Uses custom `.ddev/.env.*` files to expand env variables in `docker-compose.*.yaml` services.
* The variables are also expanded to `pre_install_actions` and `post_install_actions` in `install.yaml` for add-ons.
* The variables are passed to the respective service containers, i.e. everything from `.env.redis` will go to the `redis` service.
* `.env.*` files can override values that are already defined in the service, for example, `.ddev/docker-compose.busybox.yaml`:

	```yaml
	services:
	  busybox:
	    environment:
	      - THIS_VARIABLE_CAN_BE_CHANGED_FROM_ENV=some-value
	```
	And the value from `.ddev/.env.busybox` will be available in the `busybox` container:
	```dotenv
    THIS_VARIABLE_CAN_BE_CHANGED_FROM_ENV=new-value
	```

## Manual Testing Instructions

1. Check updated `ddev/ddev-redis` add-on:

- https://github.com/ddev/ddev-redis/pull/24

```
ddev add-on get https://github.com/ddev/ddev-redis/tarball/20240717_stasadev_optional_version
ddev dotenv set .ddev/.env.redis --redis-tag=7 # the order of commands is not important in this case
cat .ddev/.env.redis # REDIS_TAG="7"
ddev dotenv get .ddev/.env.redis --redis-tag # REDIS_TAG="7"
ddev restart
ddev redis-cli INFO | grep redis_version # redis_version:7.4.0
# the command below will take into account the env from .ddev/.env.redis
ddev add-on get https://github.com/ddev/ddev-redis/tarball/20240717_stasadev_optional_version
```

2. Check updated `rfay/ddev-php-patch-build` add-on:

- https://github.com/rfay/ddev-php-patch-build/pull/2

```
ddev dotenv set .ddev/.env.php-patch-build --static-php-version=8.0.10 # the order of commands is important in this case, because env is used in post_install_actions
ddev add-on get https://github.com/stasadev/ddev-php-patch-build/tarball/20240717_stasadev_optional_version
cat .ddev/.env.php-patch-build # STATIC_PHP_VERSION="8.0.10"
ddev dotenv get .ddev/.env.php-patch-build --static-php-version # STATIC_PHP_VERSION="8.0.10"
ddev restart
# the command below will take into account the env from .ddev/.env.php-patch-build
ddev add-on get https://github.com/stasadev/ddev-php-patch-build/tarball/20240717_stasadev_optional_version
```

## Automated Testing Overview



## Release/Deployment Notes
